common/src/main/java/com/wynntils/features/combat/AutoAttackFeature.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright © Wynntils 2024.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features.combat;
+
+import com.wynntils.core.components.Models;
+import com.wynntils.core.consumers.features.Feature;
+import com.wynntils.core.persisted.config.Category;
+import com.wynntils.core.persisted.config.ConfigCategory;
+import com.wynntils.core.text.StyledText;
+import com.wynntils.mc.event.ArmSwingEvent;
+import com.wynntils.mc.event.ArmSwingEvent.ArmSwingContext;
+import com.wynntils.mc.event.ChangeCarriedItemEvent;
+import com.wynntils.mc.event.SetSlotEvent;
+import com.wynntils.mc.event.TickEvent;
+import com.wynntils.mc.event.UseItemEvent;
+import com.wynntils.models.character.type.ClassType;
+import com.wynntils.models.spells.event.SpellEvent;
+import com.wynntils.models.spells.type.SpellDirection;
+import com.wynntils.utils.mc.LoreUtils;
+import com.wynntils.utils.mc.McUtils;
+import com.wynntils.utils.wynn.ItemUtils;
+import net.minecraft.client.player.LocalPlayer;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.item.ItemStack;
+import net.neoforged.bus.api.SubscribeEvent;
+
+@ConfigCategory(Category.COMBAT)
+public class AutoAttackFeature extends Feature {
+    private static final int TICKS_PER_ATTACK = 2;
+
+    private enum WeaponStatus {
+        UNKNOWN,
+        NOT_USABLE,
+        USABLE
+    }
+
+    private WeaponStatus weaponStatus = WeaponStatus.UNKNOWN;
+    private int lastSelectedSlot;
+    private boolean preventWrongCast = false;
+
+    @SubscribeEvent
+    public void onChangeCarriedItemEvent(ChangeCarriedItemEvent event) {
+        weaponStatus = WeaponStatus.UNKNOWN;
+    }
+
+    @SubscribeEvent
+    public void onSetSlotEvent(SetSlotEvent.Post event) {
+        if (event.getSlot() == McUtils.inventory().selected) {
+            weaponStatus = WeaponStatus.UNKNOWN;
+        }
+    }
+
+    @SubscribeEvent
+    public void onSwing(ArmSwingEvent event) {
+        if (event.getActionContext() != ArmSwingContext.ATTACK_OR_START_BREAKING_BLOCK) return;
+        if (event.getHand() != InteractionHand.MAIN_HAND) return;
+        if (Models.Character.getClassType() != ClassType.ARCHER) return;
+        lastSelectedSlot = McUtils.inventory().selected;
+        preventWrongCast = true;
+    }
+
+    @SubscribeEvent
+    public void onUseItem(UseItemEvent event) {
+        if (Models.Character.getClassType() == ClassType.ARCHER) return;
+        lastSelectedSlot = McUtils.inventory().selected;
+        preventWrongCast = true;
+    }
```

Comment:
I think you should be able to combine these using the `SpellEvent.Partial` event, then you won't need to do any archer checks.

Also, if you start casting a spell but do not finish then the feature will not work until you complete a spell so if a partial spell is in progress you should set `preventWrongCast` back to false after the spell inputs have expired, I believe it lasts for 3 seconds

Reply:
And the same goes for a failed spell actually, so use `SpellEvent.Failed` for that

Reply:
`SpellEvent` is based on a packet being received from the server and therefore not really usable for this use case.
`SpellEvent.Completed` is used as, without good network prediction, it's the only way to check for spell completion (even though it can drop a small amount of DPS due to network delay). `SpellEvent.Completed` is also called for failed casts as well.

Added a timeout for the spell cast, though it can cast a spell in some rare cases (i.e. user is holding down left click while the spell cast is expiring).